tests/test_main.py

```diff
@@ -0,0 +1,318 @@
+    runner = CliRunner()
+    requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                       json=post_matching_helper)
+
+    # copy-course <SOURCE_ID> <COURSE_NAME> <COURSE_SHORT_NAME>
+    # <COURSE_CATEGORY_ID>
+    result = runner.invoke(cli, ['copy-course', '3', 'Algebra3', 'ALG3', '1'],
+                           env={'MOODLE_URL': TEST_MOODLE_URL,
+                                'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+    assert result.exit_code == 0
+
+
+def test_course_bulk_csv(requests_mock, tmp_path):
+    test_json = {'foo': 'bar'}
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=test_json)
+        # course-bulk-csv  <OUTPUT_CSV>
+        result = runner.invoke(cli, ['course-bulk-csv', 'output.csv'],
+                               env={'MOODLE_URL': TEST_MOODLE_URL,
+                                    'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+        assert result.exit_code == 0
+        assert os.stat("output.csv").st_size != 0
+
+
+def test_course_bulk_setup(requests_mock, tmp_path):
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=get_matching_helper)
+        requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                           json=post_matching_helper)
+
+        f = open('test.csv', 'w')
+        writer = csv.DictWriter(
+                f,
+                utils.course_bulk_input_csv_fieldnames()
+            )
+        writer.writeheader()
+        writer.writerows([{utils.CSV_INST_FNAME: 'Tom',
+                           utils.CSV_INST_LNAME: 'Michaels',
+                           utils.CSV_INST_EMAIL: 'tommichaels@gmail.com',
+                           utils.CSV_INST_AUTH: 'manual',
+                           utils.CSV_COURSE_NAME: 'Algebra',
+                           utils.CSV_COURSE_SHORTNAME: 'a1',
+                           utils.CSV_COURSE_CATEGORY: 'MISC'},
+                          {utils.CSV_INST_FNAME: 'Christy',
+                           utils.CSV_INST_LNAME: 'Yolanda',
+                           utils.CSV_INST_EMAIL: 'christyyolanda@gmail.com',
+                           utils.CSV_INST_AUTH: 'manual',
+                           utils.CSV_COURSE_NAME: 'English',
+                           utils.CSV_COURSE_SHORTNAME: 'e1',
+                           utils.CSV_COURSE_CATEGORY: 'MISC'}])
+        f.close()
+
+        result = runner.invoke(cli, ['course-bulk-setup',
+                                     '2', 'test.csv', 'output.csv'],
+                               env={'MOODLE_URL': TEST_MOODLE_URL,
+                                    'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+
+        assert result.exit_code == 0
+        assert os.stat("output.csv").st_size != 0
```

Comment:
In this specific test case, it might be worth parsing out the output CSV with `csv.DictReader` and asserting it has the expected new columns `CSV_COURSE_ID,  CSV_COURSE_ENROLMENT_URL, CSV_COURSE_ENROLMENT_KEY` and number of data rows.